### PR TITLE
Implement OPTIONS support

### LINKED ISSUE #17: Implement HTTP `OPTION` support
Implement support for HTTP `OPTION` requests, should pass cob_spec's [OPTION](https://github.com/8thlight/cob_spec/blob/master/FitNesseRoot/HttpTestSuite/FileServerTestSuite/SimpleOption/content.txt) test.

### PR DESCRIPTION
**Pull request by bgerstle**
> Resolves #17

Required implementing header serialization, as well as changes to `FileHttpResponder` request handling.

### CODE REVIEW

**Comment by bgerstle** [comment]
Tested while running the server in the repo:

```
➭ curl -i -v -X OPTIONS localhost:8080/src
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> OPTIONS /src HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 200 Allow: OPTIONS,HEAD
HTTP/1.1 200 Allow: OPTIONS,HEAD
```

**Comment by bgerstle** [comment]
@trptcolin @davetorre perhaps I can't get by with my rough guesses of what ABNF means 😅 . Aside my issues interpreting the spec, I didn't quite test-drive this branch—despite my intention to do so.

**Comment by bgerstle** [comment]
Also, the `Response` class is getting rather large. I put a lot of work into refactoring `Request`, and I'm starting to wonder if `Response` would benefit from similar treatment.

**Comment by bgerstle** [comment]
Also, RFC7321 section on the OPTIONS request (4.3.7) says:

>  A server MUST generate a Content-Length field with a value of "0" if no payload body is to be sent in the response.

But cob_spec doesn't require this and it seems both Apache and Python's `urllib` are permissive enough to handle responses without it. That said, I might add it anyway.

**Comment by davetorre** [comment]
> Ok, cob_spec passes with the exception of the `/logs` special case, which seems related to some other "feature" I haven't implemented.
> 
> image

Is this how I reply?

File: src/main/java/com/eighthlight/fabrial/http/FileHttpResponder.java [hunk 1]
```diff
@@ -16,19 +18,37 @@ public FileHttpResponder(DataSource dataSource) {
...
-    // ideally this would have its own namespace, like `/static/*`
+    // ideally this would have its own namespace, like `/static/*` or `/files/*`
     return true;
   }
 
   @Override
   public Response getResponse(Request request) {
     boolean fileExists = dataSource.fileExistsAtPath(Paths.get(request.uri.getPath()));
-    if (!fileExists) {
+    if (!fileExists && request.method != Method.OPTIONS) {
```
[the remark this answers is not in the document]
Comment (davetorre) [reply src/main/java/com/eighthlight/fabrial/http/FileHttpResponder.java]: Yeah that's interesting

File: src/test/java/com/eighthlight/fabrial/test/http/HttpHeaderWriterSerializationTests.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  void writeMultipleFields() throws IOException {
+    try (var os = new ByteArrayOutputStream()) {
+      var writer = new HttpHeaderWriter(os);
+      writer.writeFields(Map.of(
+          "Allow", "GET,HEAD,OPTIONS",
+          "Content-Length", "0"));
+      var expectedAllow = "Allow: GET,HEAD,OPTIONS ";
+      var expectedContentLength = "Content-Length: 0 ";
+      var headerLines = Arrays.asList(os.toString().split("\r\n"));
+      assertThat(headerLines, containsInAnyOrder(expectedAllow, expectedContentLength));
```
Comment (davetorre) [inline src/test/java/com/eighthlight/fabrial/test/http/HttpHeaderWriterSerializationTests.java]: Small thing, I usually find it easier to read through unit tests when there's a blank lines between setup/action and between action/assertion. So here maybe have `writer.writeFields(...)` have a newline before and after.

**Comment by davetorre** [comment]
As for the Response class, the size doesn't bother. Maybe once you start writing the response body. You'll see in the S from SOLID that Uncle Bob talks about responsibilities as meaning reasons to change, but I have an easier time thinking in terms of what a class's jobs are. Like Response currently is responsible for writing the response to an OutputStream but it also does a little work preparing the data to be written. Maybe the preparation step is a second responsibility, but it doesn't seem like too much for Response to handle to me at this point.

File: src/test/java/com/eighthlight/fabrial/test/http/HttpHeaderWriterSerializationTests.java (hunk 2 above)
Comment (bgerstle) [reply src/test/java/com/eighthlight/fabrial/test/http/HttpHeaderWriterSerializationTests.java]: I prefer to refactor into expressive helper methods where it makes sense, but where that fails, I've also found newlines helpful in distinguishing between the different "phases".

**Comment by bgerstle** [comment]
I realize now where I messed up. The `status-line` part of the HTTP message spec includes a newline, which I had overlooked before.

**Pull request merged**